### PR TITLE
Add support for concurrent collection when size is greater than zero

### PR DESCRIPTION
**Pull request by javanna**
## Query phase support for concurrency when size > 0

The query phase supports so far concurrency only when size is set to 0 (despite inter-segment concurrency is not yet enabled in the query phase). This commit adds support for concurrent collector managers to the query phase when size is greater than zero (hits are requested), including when scroll is used. The only exception is field collapsing, which relies on custom collectors for which we don't have yet a corresponding collector manager. 

Note: this commit does not enable concurrent collection for the query phase. The entire query phase still collects against a single slice, sequentially segment by segment.

## Collector managers refactoring / suppliers removal

As part of this change, especially figuring out how to leverage the lucene collector managers for top score doc and top field doc collectors, many simplifications were made, that are explained as follows.

We currently create multiple collector managers for the query phase: one for the top docs collection, one for aggs, and both of them are wrapped into a query phase collector manager. Additionally, each one of those is wrapped into a ProfileCollectorManager if profiling is enabled. While this works, the 1:1 ratio between collectors and collector managers is not necessary and makes things quite hard to follow. 

Also, TopDocsCollectorManagerFactory used to be a factory to create the right top docs collector based on the requested features, and it is now a factory of collector managers. We can simplify things and remove one level of indirection by converting it into a collector manager directly. Also, we use some suppliers to keep track of which collector provides what part of the search response (e.g. total hits, top docs, max score). This is because of the collectors wrapping: before the introduction of collector managers, it was difficult to extract the appropriate info from each collector after calling search, problem that was solved using suppliers. Collector managers are a good fit to replace these indirections, as they can return a value, which is the result of the reduction of the results obtained by each slice after collection.

This commit folds all of the different collector managers that we rely on for the query phase into a single one (called QueryPhaseCollectorManager): its newCollector method will create the appropriate collectors depending on the requested features (a bit like TopDocsCollectorManagerFactory used to do), but it will take care of the creation of all the collectors involved, depending on feature requested (size, scroll, collapse) or whether profiling was enabled. This allows us to have a single base collector manager that includes all the logic around creating collectors and reducing their results, as well as returning a concrete result which includes the top docs, terminated after flag, profiling tree if profiling was enabled which entirely removes the need for suppliers in multiple places.

There are no suppliers needed for total hits and top docs, as they are returned directly by the search method. There is no need for collector managers to hold state that is only set once reduce is called. Profiling was greatly simplified as well to no longer rely on suppliers for the profiler results. Instead, the collector results can simply be set to the query profiler itself after search returns. Additionally, ProfileCollectorManager used to have handling for children collectors for top docs and aggs, which was very hard to follow. This is no longer required because QueryPhaseCollectorManager takes care of it all (wrapping top docs, aggs, as well as the top level collector when profiling is enabled) and removes the need for wrapping of collector managers.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java [hunk 1]
```diff
@@ -13,73 +13,31 @@
...
 public final class ProfileCollectorManager<T> implements CollectorManager<InternalProfileCollector, T> {
-
     private final CollectorManager<? extends Collector, T> collectorManager;
     private final String reason;
-    private final ProfileCollectorManager<?> topDocsSubCollectorManager;
-    private final ProfileCollectorManager<?> aggsSubCollectorManager;
-    // this is a bit of a hack: it allows us to retrieve the last collector that newCollector has returned for sub-collector managers,
-    // so that we can provide them to InternalProfileCollector's constructor as children. This is fine as newCollector does not get called
-    // concurrently, but rather in advance before parallelizing the collection
-    private InternalProfileCollector profileCollector;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java:37]: query phase is the only place where we have a top level collector wrapped with a profile collector, as well as its inner collectors (for top docs and aggs) both wrapped with a profile collector. Thanks to removing collector manager wrapping, this is now directly handled in QueryPhaseCollectorManager within newCollector, and this complicated bit can be removed from ProfileCollectorManager, which is only used by the dfs phase. There we do need to wrap the collector manager because we rely on the lucene top docs collector managers.

File: server/src/main/java/org/elasticsearch/search/profile/query/QueryProfiler.java [hunk 2]
```diff
@@ -28,20 +27,20 @@
 public final class QueryProfiler extends AbstractProfiler<QueryProfileBreakdown, Query> {
 
     /**
-     * The root Collector used in the search
+     * The root CollectorResult used in the search
      */
-    private Supplier<CollectorResult> collectorResultSupplier;
+    private CollectorResult collectorResult;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/profile/query/QueryProfiler.java:32]: This supplier was really only needed because it was difficult to keep track of the collector (or manager) that provides us with the top level collector result tree. As that is now part of the result returned by the search method, it is very simple to set the result after search has been called which allows us to remove the need for the supplier.

File: server/src/main/java/org/elasticsearch/search/query/PartialHitCountCollector.java [hunk 3]
```diff
@@ -81,43 +80,4 @@ boolean isThresholdReached() {
             return numCollected.getAcquire() >= totalHitsThreshold;
         }
     }
-
-    static class CollectorManager implements org.apache.lucene.search.CollectorManager<PartialHitCountCollector, Void> {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/PartialHitCountCollector.java:85]: This functionality is now folded into QueryPhaseCollectorManager.EmptyHits. This small collector manager has been moved to PartialHitCountCollectorTests to allow for testing this collector in concurrent scenarios (as opposed to testing the collector only in sequential collection scenarios).

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 4]
```diff
@@ -214,38 +194,34 @@ static void addCollectorsAndSearch(SearchContext searchContext) throws QueryPhas
...
-            QueryPhaseCollector.CollectorManager queryPhaseCollectorManager = QueryPhaseCollector.createManager(
-                topDocsCollectorManager,
+            CollectorManager<Collector, QueryPhaseResult> collectorManager = QueryPhaseCollectorManager.createQueryPhaseCollectorManager(
                 postFilterWeight,
-                searchContext.terminateAfter(),
                 aggsCollectorManager,
-                searchContext.minimumScore()
+                searchContext,
+                hasFilterCollector
             );
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:202]: All of the wrapping is no longer necessary. Collectors are still wrapped when necessary, but there's a single collector manager taking care of that, including profiling.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhase.java [hunk 5]
```diff
@@ -266,27 +242,6 @@ static void addCollectorsAndSearch(SearchContext searchContext) throws QueryPhas
...
-            assert timeoutSet : "TimeExceededException thrown even though timeout wasn't set";
-            if (searchContext.request().allowPartialSearchResults() == false) {
-                throw new QueryPhaseExecutionException(searchContext.shardTarget(), "Time exceeded");
-            }
-            queryResult.searchTimedOut(true);
-        }
-        if (searchContext.terminateAfter() != SearchContext.DEFAULT_TERMINATE_AFTER && queryResult.terminatedEarly() == null) {
-            queryResult.terminatedEarly(false);
-        }
-    }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:288]: rather than having this method return a QueryPhaseResult, I folded it into its caller method, as that was the only caller for it.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java [hunk 6]
```diff
@@ -52,10 +49,6 @@ public final class QueryPhaseCollector implements Collector {
     private final boolean cacheScores;
     private boolean terminatedAfter = false;
 
-    QueryPhaseCollector(Collector topDocsCollector, Weight postFilterWeight, int terminateAfter, Collector aggsCollector, Float minScore) {
-        this(topDocsCollector, postFilterWeight, resolveTerminateAfterChecker(terminateAfter), aggsCollector, minScore);
-    }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java:57]: Removed this as it was needed only for sequential scenarios. In fact the terminate after checker needs to be shared across the different collectors that the same instance of the collector manager creates.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java [hunk 7]
```diff
@@ -389,77 +374,6 @@ boolean incrementHitCountAndCheckThreshold() {
         }
     };
 
-    /**
-     * {@link org.apache.lucene.search.CollectorManager} implementation based on {@link QueryPhaseCollector}.
-     * Wraps two {@link org.apache.lucene.search.CollectorManager}s: one required for top docs collection, and another one optional for
-     * aggs collection. Applies terminate_after consistently across the different collectors by sharing an atomic counter of collected docs.
-     */
-    static class CollectorManager implements org.apache.lucene.search.CollectorManager<QueryPhaseCollector, Void> {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java:397]: This functionality is now part of QueryPhaseCollectorManager. This collector manager has been simplified and moved to QueryPhaseCollectorTests to allow for testing QueryPhaseCollector in concurrent scenarios, as opposed to testing it only in sequential scenarios.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+                (InternalProfileCollector) aggsCollector
+            );
+        }
+        return queryPhaseCollector;
+    }
+
+    protected abstract Collector newTopDocsCollector() throws IOException;
+
+    @Override
+    public final QueryPhaseResult reduce(Collection<Collector> collectors) throws IOException {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:140]: The casting required in this method is perhaps the only cons I see around removing collector managers wrapping, as you can have many different types of collectors created, that need to be reduced. On the other hand, the structure is quite predictable, and it depends on the logic that newCollector includes above. I think that is reasonable and the pros of removing the wrapping outweigh the need for casting collectors in this method.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+        }
+        return Float.NaN;
+    }
+
+    /**
+     * Returns query total hit count if the <code>query</code> is a {@link MatchAllDocsQuery}
+     * or a {@link TermQuery} and the <code>reader</code> has no deletions,
+     * -1 otherwise.
+     */
+    static int shortcutTotalHitCount(IndexReader reader, Query query) throws IOException {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:622]: this is nothing new, it was copied with no changes form TopDocsCollectorManagerFactory. Nothing to look at.

File: server/src/main/java/org/elasticsearch/search/query/SingleThreadCollectorManager.java [hunk 10]
```diff
@@ -1,39 +0,0 @@
...
-package org.elasticsearch.search.query;
-
-import org.apache.lucene.search.Collector;
-import org.apache.lucene.search.CollectorManager;
-
-import java.io.IOException;
-import java.util.Collection;
-
-/** {@link CollectorManager} that runs in a non-concurrent search. */
-public class SingleThreadCollectorManager implements CollectorManager<Collector, Void> {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/SingleThreadCollectorManager.java:18]: no more need for single threaded collector managers. The only place where we have that need is when field collapsing is used, for which we assert that we do single slices collection directly in QueryPhaseCollectorManager#forCollapsing

File: server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorManagerFactory.java [hunk 11]
```diff
@@ -1,527 +0,0 @@
...
-                return TopScoreDocCollector.create(numHits, searchAfter, hitCountThreshold);
-            } else {
-                return TopFieldCollector.create(sortAndFormats.sort, numHits, (FieldDoc) searchAfter, hitCountThreshold);
-            }
-        }
-
-        private final Collector collector;
-        private final Supplier<TotalHits> totalHitsSupplier;
-        private final Supplier<TopDocs> topDocsSupplier;
-        private final Supplier<Float> maxScoreSupplier;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorManagerFactory.java:217]: These three suppliers are no longer needed, which feels great. They were used to set the right values within TopDocsAndMaxScore, which they were coming from different places depending on input parameters. It was previously difficult to trace which collector was providing them and how, due to the wrapping in different places. Now, we can have different subclasses of QueryPhaseCollectorManager (similar to what we previously has in TopDocsCollectorManagerFactory) that provide values depending on their logic, and those values become automatically part of the end result that the collector manager reduces which the search method returns.

This is the main reason to refactor the query phase like this PR does, as part of introducing support for concurrent collection to it. It felt really bad to convert to concurrent collector manager and keep the suppliers around.

File: server/src/test/java/org/elasticsearch/search/profile/query/ProfileCollectorManagerTests.java [hunk 12]
```diff
@@ -142,110 +137,4 @@ public void testManagerWithSearcher() throws IOException {
             assertTrue(result.getTime() > 0);
         }
     }
-
-    public void testManagerWithChildren() throws IOException {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/search/profile/query/ProfileCollectorManagerTests.java:146]: These were testing support for children collectors, which ProfileCollectorManager no longer support

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java [hunk 13]
```diff
@@ -98,176 +99,143 @@ public static void afterClass() throws Exception {
...
 
     public void testTopDocsOnly() throws IOException {
         {
             CollectorManager<TopScoreDocCollector, TopDocs> topScoreDocManager = TopScoreDocCollector.createSharedManager(1, null, 1000);
-            CollectorManagerAdapter<TopScoreDocCollector, TopDocs> topScoreDocAdapter = new CollectorManagerAdapter<>(topScoreDocManager);
-            QueryPhaseCollector.CollectorManager manager = QueryPhaseCollector.createManager(topScoreDocAdapter, null, 0, null, null);
-            searcher.search(new MatchAllDocsQuery(), manager);
-            assertFalse(manager.isTerminatedAfter());
-            assertEquals(numDocs, topScoreDocAdapter.getResult().totalHits.value);
+            CollectorManager<QueryPhaseCollector, Result<TopDocs, Void>> manager = createCollectorManager(
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:112]: Small suggestion: these might be easier to read with `var`

**Review by romseygeek — APPROVED** [review]
LGTM - this is a really nice cleanup/simplification.  I think profiling becomes a lot simpler if/when we get something like LeafCollector.finish() which could remove a lot of the extra casting, but at least for now it's all in one method.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+            terminateAfterChecker,
+            aggsCollector,
+            minScore
+        );
+        if (profile) {
+            if (aggsCollector == null) {
+                return new InternalProfileCollector(
+                    queryPhaseCollector,
+                    REASON_SEARCH_QUERY_PHASE,
+                    (InternalProfileCollector) topDocsCollector
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:124]: that's a good idea. Will do that.

File: server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java (hunk 13 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:112]: interesting: certainly less typing. Though I need to replace Void with Object somehow. It is less heavy, but is it easier to read given that it's not entirely clear which type I get?

Comment (javanna) [reply server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorTests.java:112]: I haven't made this change for now. Can do that as a follow-up perhaps.

**Review by JVerwolf — COMMENTED** [review]
I wasn't able to do a full round today, just left a single comment.

**Review by jdconrad — APPROVED** [review]
LGTM! This PR really illustrates how invasive profiling is once again, but the amount of work reduced around child profilers is fantastic.

File: server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java (hunk 8 above)
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:146]: It's not ideal, but I agree with the trade-off here to significantly reduce the abstractions.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java:146]: This also comes from the collector manager design: you create collector, and get the same ones as arguments to reduce. It would probably make things easier to just keep track of the collectors that were created and just reduce those, then you don't need casting etc. that feels though not very aligned with the Lucene design. Perhaps this is something that we could look at simplifying in Lucene.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 15]
```diff
@@ -1104,7 +1104,8 @@ static boolean supportsParallelCollection(ResultsType resultsType, SearchSourceB
         /*
         TODO uncomment this block to enable inter-segment concurrency for the query phase
         if (resultsType == ResultsType.QUERY) {
-            return source == null || source.aggregations() == null || source.aggregations().supportsParallelCollection();
+            return source == null || source.collapse() == null || source.aggregations() == null
+            || source.aggregations().supportsParallelCollection();
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1108]: Perhaps yes, but it's a single situation for top docs collection compared to aggs where some of them don't support concurrency. I felt like it may be overkill to add a supportParallelCollection to SearchSourceBuilder. Maybe though that's a good idea, let me play with it.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 16]
```diff
@@ -1104,7 +1104,8 @@ static boolean supportsParallelCollection(ResultsType resultsType, SearchSourceB
         /*
         TODO uncomment this block to enable inter-segment concurrency for the query phase
```
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1105]: not an issue, but a PR that @JVerwolf is working on , see #98455

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1108]: well, this made me realize that the conditional was wrong :) and that if I move it to SearchSourceBuilder I can test it separately from enabling parallel collection in the query phase (this method in SearchService can stay unchanged).

[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Here you go: https://github.com/elastic/elasticsearch/pull/98425/commits/d239f0bac1d58fccd5828e0e6b54d79fffef7927
